src/cdomain/value/cdomains/int/congruenceDomain.ml

```diff
@@ -319,8 +319,25 @@ struct
         pretty res ;
     res
 
-  let sub ?(no_ov=false) ik x y = add ~no_ov ik x (neg ~no_ov ik y)
-
+  let sub ?(no_ov=false) ik x y =
+    let no_ov_case (c1, m1) (c2, m2) =
+      c1 -: c2, Z.gcd m1 m2
+    in
+    match (x, y) with
+    | None, None -> bot ()
+    | None, _
+    | _, None ->
+      raise (ArithmeticOnIntegerBot (Printf.sprintf "%s op %s" (show x) (show y)))
+    | Some a, Some b when no_ov ->
+      normalize ik (Some (no_ov_case a b))
+    | Some (c1, m1), Some (c2, m2) when m1 =: Z.zero && m2 =: Z.zero && not (Cil.isSigned ik) ->
+      let _, max_ik = range ik in
+      let m_ikind = max_ik +: Z.one in
+      let c = (c1 -: c2 +: m_ikind) %: m_ikind  in
+      Some(c, Z.zero)
```

Comment:
Can't we also do better in the signed constants case? Namely, when `c1-c2` is within range of the type? Or are we guaranteed to be passed the `no_ov` flag in these case?

Reply:
I added a dedicated treatment for that case.